source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,523 @@
+    /// No Response
+    None,
+    /// Response with 48-bit length
+    Short,
+    /// Response with 48-bit length + check *busy* after response
+    ShortWithBusySignal,
+    /// Response with 136-bit length
+    Long,
+}
+
+/// Response returned by the card, can be short or long, depending on command
+#[must_use]
+pub enum Response {
+    /// The 32-bit value from the 48-bit response.
+    /// Potentially also includes the data buffer if this was a read command
+    Short {
+        /// The response on the command line
+        value: u32,
+        /// The received data, in case of a read command
+        data: Option<FixedVec<u8>>,
+    },
+    /// The 128-bit value from the 136-bit response
+    Long(u128),
+}
+
+/// Errors returned by the [`SdmmcService`]
+#[derive(Debug, Eq, PartialEq)]
+pub struct Error {
+    kind: ErrorKind,
+    message: Option<&'static str>,
+}
+
+/// The different types of errors that can occur
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// The service is currently busy and cannot handle the request
+    Busy,
+    /// Invalid or unexpected response was received
+    Response,
+    /// Invalid or unexpected data was received
+    Data,
+    /// The provided buffer does not meet the requirements
+    Buffer,
+    /// A timeout occurred
+    Timeout,
+    /// A different error has occurred
+    Other,
+}
+
+impl Default for Command {
+    fn default() -> Self {
+        Command {
+            index: 0,
+            argument: 0,
+            options: HardwareOptions::None,
+            kind: CommandKind::Control,
+            rsp_type: ResponseType::None,
+            rsp_crc: false,
+            buffer: None,
+        }
+    }
+}
+
+impl Error {
+    /// Create an error from a type and message
+    pub fn new(kind: ErrorKind, message: &'static str) -> Self {
+        Self {
+            kind,
+            message: Some(message),
+        }
+    }
+}
+
+impl From<ErrorKind> for Error {
+    fn from(value: ErrorKind) -> Self {
+        Self {
+            kind: value,
+            message: None,
+        }
+    }
+}
+
+impl core::fmt::Display for Error {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "SD/MMC error: ")?;
+        match self.kind {
+            ErrorKind::Busy => write!(f, "busy "),
+            ErrorKind::Response => write!(f, "response "),
+            ErrorKind::Data => write!(f, "data "),
+            ErrorKind::Buffer => write!(f, "buffer "),
+            ErrorKind::Timeout => write!(f, "timeout "),
+            ErrorKind::Other => write!(f, "other "),
+        }?;
+
+        self.message.map_or_else(|| Ok(()), |msg| msg.fmt(f))
+    }
+}
+
+impl core::fmt::Debug for Response {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Response::Short { value, data: _ } => write!(f, "{:#x}", value),
+            Response::Long(value) => write!(f, "{:#x}", value),
```

Comment:
tiny style nit, feel free to ignore me:
```suggestion
            Response::Short { value, data: _ } => write!(f, "{value:#x}"),
            Response::Long(value) => write!(f, "{value:#x}"),
```